File: frontend/src/types/evaluation.ts
export interface RubricCriterionEvaluation {
  criterion_id: string;
  criterion_text: string;
  max_marks: number;
  awarded_marks: number;
  reasoning: string;
}

export interface EvaluationItem {
  id: string;
  evaluation_id: string;
  question_paper_item_id: string;
  extracted_answer_id?: string;
  question_number: string;
  maximum_marks: number;
  ai_marks: number;
  final_marks?: number;
  keyword_score: number;
  concept_score: number;
  semantic_score: number;
  pattern_score: number;
  rubric_score: number;
  ai_score: number;
  confidence: number;
  matched_keywords?: any;
  missing_keywords?: any;
  matched_concepts?: any;
  missing_concepts?: any;
  strengths?: any;
  missing_points?: any;
  evaluation_explanation?: string;
  rubric_evaluations?: RubricCriterionEvaluation[];
  requires_faculty_review: boolean;
  review_reason?: string;
  review_status?: string; // 'PENDING' | 'ACCEPTED' | 'OVERRIDDEN' | 'RE_EVALUATED'
  override_reason?: string;
  faculty_comment?: string;
  reviewed_by?: string;
  reviewed_at?: string;
}

export interface Evaluation {
  id: string;
  examination_id: string;
  answer_paper_id: string;
  student_id: string;
  version: number;
  evaluator_type: 'AI' | 'FACULTY' | 'HYBRID';
  status: 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'UNDER_REVIEW' | 'APPROVED' | 'FINALIZED' | 'REOPENED' | 'REJECTED';
  requires_review: boolean;
  total_ai_marks: number;
  total_final_marks?: number;
  overall_confidence: number;
  started_at?: string;
  completed_at?: string;
  reviewed_at?: string;
  reviewed_by?: string;
  approved_by?: string;
  approved_at?: string;
  finalized_by?: string;
  finalized_at?: string;
  review_notes?: string;
  finalization_notes?: string;
  items: EvaluationItem[];
}

export interface OverrideEvaluationItemRequest {
  final_marks: number;
  reason: string;
  comment?: string;
}

export interface AcceptAIRequest {
  comment?: string;
}

export interface BulkAcceptRequest {
  confidence_threshold?: number;
}

export interface ApproveEvaluationRequest {
  notes?: string;
}

export interface FinalizeEvaluationRequest {
  notes?: string;
}

export interface ReopenEvaluationRequest {
  reason: string;
}

export interface FacultyReviewRecord {
  id: string;
  evaluation_id: string;
  evaluation_item_id?: string;
  faculty_id: string;
  original_ai_marks?: number;
  revised_marks: number;
  comment?: string;
  reason?: string;
  action: string;
  reviewed_at?: string;
}

File: backend/app/schemas/evaluation.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from backend.app.db.models.enums import EvaluatorTypeEnum, EvaluationStatusEnum

class EvaluationItemResponse(BaseModel):
    id: str
    evaluation_id: str
    question_paper_item_id: str
    extracted_answer_id: Optional[str] = None
    maximum_marks: float
    keyword_score: Optional[float] = None
    concept_score: Optional[float] = None
    semantic_score: Optional[float] = None
    pattern_score: Optional[float] = None
    rubric_score: Optional[float] = None
    ai_score: Optional[float] = None
    ai_marks: Optional[float] = None
    final_marks: Optional[float] = None
    confidence: Optional[float] = None
    matched_keywords: Optional[Dict[str, Any]] = None
    missing_keywords: Optional[Dict[str, Any]] = None
    matched_concepts: Optional[Dict[str, Any]] = None
    missing_concepts: Optional[Dict[str, Any]] = None
    strengths: Optional[Dict[str, Any]] = None
    missing_points: Optional[Dict[str, Any]] = None
    criterion_scores: Optional[Dict[str, Any]] = None
    requires_faculty_review: bool
    review_reason: Optional[str] = None
    prompt_version: Optional[str] = None
    evaluation_explanation: Optional[str] = None
    review_status: Optional[str] = "PENDING"
    override_reason: Optional[str] = None
    faculty_comment: Optional[str] = None
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[str] = None

    class Config:
        from_attributes = True


class EvaluationResponse(BaseModel):
    id: str
    examination_id: str
    answer_paper_id: str
    student_id: str
    version: int
    evaluator_type: EvaluatorTypeEnum
    status: EvaluationStatusEnum
    requires_review: bool
    total_ai_marks: Optional[float] = None
    total_final_marks: Optional[float] = None
    overall_confidence: Optional[float] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    reviewed_at: Optional[str] = None
    reviewed_by: Optional[str] = None
    approved_by: Optional[str] = None
    approved_at: Optional[str] = None
    finalized_by: Optional[str] = None
    finalized_at: Optional[str] = None
    review_notes: Optional[str] = None
    finalization_notes: Optional[str] = None
    items: List[EvaluationItemResponse] = []

    class Config:
        from_attributes = True


class EvaluationSummaryResponse(BaseModel):
    evaluation_id: str
    answer_paper_id: str
    student_id: str
    version: int
    status: str
    total_ai_marks: Optional[float] = None
    total_final_marks: Optional[float] = None
    overall_confidence: Optional[float] = None
    requires_review: bool
    total_items: int
    reviewed_items: int


class ReevaluateItemRequest(BaseModel):
    reason: Optional[str] = Field(None, description="Faculty reason for item re-evaluation")


class AcceptAIRequest(BaseModel):
    comment: Optional[str] = Field(None, description="Optional faculty comment when accepting AI mark")


class OverrideMarksRequest(BaseModel):
    final_marks: float = Field(..., description="Revised final marks awarded by faculty")
    reason: str = Field(..., min_length=10, description="Mandatory reason for mark override (min 10 chars)")
    comment: Optional[str] = Field(None, description="Optional additional faculty comment")


class RequestReevaluationRequest(BaseModel):
    reason: str = Field(..., min_length=5, description="Reason for requesting AI re-evaluation")


class BulkAcceptRequest(BaseModel):
    confidence_threshold: float = Field(0.65, ge=0.0, le=1.0, description="Minimum confidence threshold for bulk acceptance")


class ApproveEvaluationRequest(BaseModel):
    notes: Optional[str] = Field(None, description="Optional approval notes")


class FinalizeEvaluationRequest(BaseModel):
    notes: Optional[str] = Field(None, description="Optional finalization notes")


class ReopenEvaluationRequest(BaseModel):
    reason: str = Field(..., min_length=10, description="Mandatory administrative reason for reopening finalized evaluation")


class FacultyReviewResponse(BaseModel):
    id: str
    evaluation_id: str
    evaluation_item_id: Optional[str] = None
    faculty_id: str
    original_ai_marks: Optional[float] = None
    revised_marks: float
    comment: Optional[str] = None
    reason: Optional[str] = None
    action: str
    reviewed_at: Optional[str] = None

    class Config:
        from_attributes = True
